csharp/src/Ice/BZip2.cs

```diff
@@ -301,49 +338,91 @@ private static string GetBZ2Error(int error)
         public static Ice.OutputStream? Compress(Ice.OutputStream stream, int headerSize, int compressionLevel)
         {
             Debug.Assert(Supported());
-            //
+            var bzStream = new BZStream();
+            bzStream.BzAlloc = IntPtr.Zero;
+            bzStream.BzFree = IntPtr.Zero;
+            bzStream.Opaque = IntPtr.Zero;
+
+            _compressInit(ref bzStream, compressionLevel, 0, 0);
             // Compress the message body, but not the header.
-            //
             int uncompressedLen = stream.Size - headerSize;
 
-            // TODO: Avoid copy the data using bzip2 low level API, feeding segments instead of allocating
-            // a new array to hold the whole input.
-            byte[] data = stream.GetBytes(headerSize, uncompressedLen);
+            byte[] compressed = new byte[(int)((uncompressedLen * 1.01) + 600)];
+
+            // Prevent GC from moving the byte array, this allow to take the object address
+            // and pass it to bzip2 calls.
+            var compressedHandle = GCHandle.Alloc(compressed, GCHandleType.Pinned);
+            bzStream.NextOut = compressedHandle.AddrOfPinnedObject();
+            bzStream.AvailOut = (uint) compressed.Length;
+            System.Collections.Generic.IList<ArraySegment<byte>> data = stream.GetUnderlyingBuffer();
+
+            int rc = BzOk;
+            for (int i = 0; i < data.Count; ++i)
+            {
+                ArraySegment<byte> segment = data[i];
+                var outHandle = GCHandle.Alloc(segment.Array, GCHandleType.Pinned);
+
+                if (i == 0)
+                {
+                    bzStream.NextIn = outHandle.AddrOfPinnedObject() + headerSize;
+                    bzStream.AvailIn = (uint)(segment.Count - headerSize);
+                }
+                else
+                {
+                    bzStream.NextIn = outHandle.AddrOfPinnedObject();
+                    bzStream.AvailIn = (uint)segment.Count;
+                }
+                Debug.Assert(bzStream.AvailIn > 0);
+
+                do
+                {
+                    rc = _compress(ref bzStream, BzRun);
+                }
+                while (rc == BzRunOk && bzStream.AvailIn > 0);
 
-            int compressedLen = (int)((uncompressedLen * 1.01) + 600);
-            byte[] compressed = new byte[compressedLen];
+                if (rc == BzRunOk && i + 1 == data.Count)
```

Comment:
You mean the call to compress with BzFinish action? I put it here because at this point bzip can still read from NextIn, and the outHandle that ensure that the array segment is still in scope is free at the end of the loop

Reply:
Fixed, NextIn is not read after BzRun action finish so it is fine to release the handle here and move this out